ballerina/query_complexity_validator_visitor.bal

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2024, WSO2 LLC. (http://www.wso2.org).
+//
+// WSO2 LLC. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+import ballerina/log;
+
+class QueryComplexityValidatorVisitor {
+    *ValidatorVisitor;
+
+    private final Engine engine;
+    private final readonly & __Schema schema;
+    private final QueryComplexityConfig queryComplexityConfig;
+    private final string operationName;
+    private final NodeModifierContext nodeModifierContext;
+    private int queryComplexity = 0;
+    private ErrorDetail[] errors = [];
+
+    isolated function init(Engine engine, readonly & __Schema schema, QueryComplexityConfig queryComplexityConfig,
+            string operationName, NodeModifierContext nodeModifierContext) {
+        self.engine = engine;
+        self.schema = schema;
+        self.queryComplexityConfig = queryComplexityConfig;
+        self.operationName = operationName;
+        self.nodeModifierContext = nodeModifierContext;
+    }
+
+    public isolated function visitDocument(parser:DocumentNode documentNode, anydata data = ()) {
+    }
+
+    public isolated function visitOperation(parser:OperationNode operationNode, anydata data = ()) {
+        if self.queryComplexityConfig.maxComplexity == 0 {
+            return;
+        }
+        __Type parentType;
+        if operationNode.getKind() == parser:OPERATION_QUERY {
+            parentType = self.schema.queryType;
+        } else if operationNode.getKind() == parser:OPERATION_MUTATION {
+            parentType = <__Type>self.schema.mutationType;
+        } else {
+            parentType = <__Type>self.schema.subscriptionType;
+        }
+        foreach parser:SelectionNode selection in operationNode.getSelections() {
+            selection.accept(self, parentType);
+        }
+        if self.queryComplexityConfig.maxComplexity < self.queryComplexity {
+            string operationName = self.operationName == parser:ANONYMOUS_OPERATION ? "" : string `${self.operationName} `;
+            string message = string `The operation ${operationName}exceeds the maximum query complexity threshold. Maximum allowed complexity: ${self.queryComplexityConfig.maxComplexity}, actual complexity: ${self.queryComplexity}`;
```

Comment:
```suggestion
            string message = string `The operation ${operationName} exceeds the maximum query complexity threshold. Maximum allowed complexity: ${self.queryComplexityConfig.maxComplexity}, actual complexity: ${self.queryComplexity}`;
```

Reply:
This was intentional to keep the formatting intact.